### ISSUE TITLE
how to train yolov8-rtdetr.yam in common

### ISSUE THREAD

**Issue by op-newman** [ultralytics/ultralytics#3597]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Dear author, hello:

I found an interesting issue in the YOLOv8 project. I found that the project included the yolov8-rtdetr.yaml configuration file. I trained my own dataset locally and modified YOLOv8. I found that the configuration file was created using the

Model=RTDETR (....) for parsing configuration files, but I found that for training,

I encountered an error while importing model. train (* * {'cfg': '... cfg. yaml'}). It seems that for rtdetr training, the built-in cfg file in the project was not imported. I don't know if this is an agency matter or if there are different cfg files for rtdetr training. As far as I know, ultra tics/yolo/cfg/default.yaml seems to be only applicable to yolov8, for yolo with a replaced delta detection head. I did not successfully train it. I don't understand how the parameters required for training are set. I would greatly appreciate it if you could provide an answer

Your most loyal reader

Bopeng

**Comment by glenn-jocher** [comment]
Just do

```bash
yolo train model=yolov8s-rtdetr.yaml data=coco8.yaml
```

**Comment by op-newman** [comment]
Thank you very much for your feedback. I have used the code locally. However, after installing UltraAnalytics, I cannot modify the code locally to achieve my goal of modifying YOLOv8. Therefore, I uninstalled UltraAnalytics and created a script in the code to run it. So commands such as yolo or RTDETR cannot be used. And I also found that RTDETR and YOLO have differences in the code for parsing cfg
this is rtdetr：
   def train(self, **kwargs):
        """
        Trains the model on a given dataset.

        Args:
            **kwargs (Any): Any number of arguments representing the training configuration.
        """
        overrides = dict(task='detect', mode='train')
        overrides.update(kwargs)
        overrides['deterministic'] = False
        if not overrides.get('data'):
            raise AttributeError("Dataset required but missing, i.e. pass 'data=coco128.yaml'")
        if overrides.get('resume'):
            overrides['resume'] = self.ckpt_path
        self.task = overrides.get('task') or self.task
        self.trainer = RTDETRTrainer(overrides=overrides)
        if not overrides.get('resume'):  # manually set model only if not resuming
            self.trainer.model = self.trainer.get_model(weights=self.model if self.ckpt else None, cfg=self.model.yaml)
            self.model = self.trainer.model
        self.trainer.train()
        # Update model and cfg after training
        if RANK in (-1, 0):
            self.model, _ = attempt_load_one_weight(str(self.trainer.best))
            self.overrides = self.model.args
            self.metrics = getattr(self.trainer.validator, 'metrics', None)  # TODO: no metrics returned by DDP
this is yolo
    def train(self, **kwargs):
        """
        Trains the model on a given dataset.

        Args:
            **kwargs (Any): Any number of arguments representing the training configuration.
        """
        self._check_is_pytorch_model()
        if self.session:  # Ultralytics HUB session
            if any(kwargs):
                LOGGER.warning('WARNING ⚠️ using HUB training arguments, ignoring local training arguments.')
            kwargs = self.session.train_args
        check_pip_update_available()
        overrides = self.overrides.copy()
        if kwargs.get('cfg'):
            LOGGER.info(f"cfg file passed. Overriding default params with {kwargs['cfg']}.")
            overrides = yaml_load(check_yaml(kwargs['cfg']))
        overrides.update(kwargs)
        overrides['mode'] = 'train'
        if not overrides.get('data'):
            raise AttributeError("Dataset required but missing, i.e. pass 'data=coco128.yaml'")
        if overrides.get('resume'):
            overrides['resume'] = self.ckpt_path
        self.task = overrides.get('task') or self.task
        self.trainer = TASK_MAP[self.task][1](overrides=overrides, _callbacks=self.callbacks)
.....
It seems that we cannot directly pass the ultralytics yolo cfg default.yaml parameter to rtdetr for parsing. May I ask which parameters rtdetr can currently parse.

**Comment by liudongcheng** [comment]
in this line "topk_ind = torch.topk(enc_outputs_scores.max(-1).values, self.num_queries, dim=1).indices.view(-1)"
error: selected index k out of range

**Comment by glenn-jocher** [comment]
@op-newman thank you for reaching out and providing the code snippets. I understand that you are encountering difficulties in training YOLOv8 with the yolov8-rtdetr.yaml configuration file and that there are differences in the code for parsing the cfg file between RTDETR and YOLO.

In the YOLOv8 code snippet you provided, the cfg file is parsed using the overrides dictionary. However, in the RTDETR code snippet, the cfg file is not directly passed as a parameter for parsing.

Based on the code snippets, it seems that the RTDETR module may not support parsing the ultra_tics/yolo/cfg/default.yaml file that is used for YOLOv8. It is possible that RTDETR has its own specific set of parameters that can be parsed.

To clarify the specific parameters that RTDETR can currently parse, I would recommend referring to the documentation or source code of RTDETR. This will provide you with detailed information on the configuration options and parameters that can be used with RTDETR.

I hope this helps clarify the situation. If you have any further questions or need additional assistance, please feel free to ask.

**Comment by liudongcheng** [comment]
收到~

**Comment by glenn-jocher** [comment]
@liudongcheng hi Bopeng,

Thank you for your interest in YOLOv8 and for reaching out with your question. 

Regarding the use of the yolov8-rtdetr.yaml configuration file for training, it seems that there is a difference between YOLOv8 and RTDETR in terms of parsing the cfg file. The YOLOv8 code snippet you provided demonstrates how the cfg file is parsed using the overrides dictionary. However, in the RTDETR code snippet, the cfg file is not directly passed as a parameter for parsing.

It appears that the RTDETR module may not currently support parsing the ultra_tics/yolo/cfg/default.yaml file that is used for YOLOv8. It is likely that RTDETR has its own specific set of parameters that can be parsed.

To obtain more specific information on the parameters that can be parsed with RTDETR, I would recommend referring to the documentation or source code of RTDETR. These resources will provide you with detailed insights into the available configuration options and parameters for training your model.

I hope this explanation clarifies the situation for you. If you have any further questions or need additional assistance, please feel free to ask.

Best regards,
Glenn Jocher

**Comment by op-newman** [comment]
> @liudongcheng hi Bopeng,
> 
> Thank you for your interest in YOLOv8 and for reaching out with your question.
> 
> Regarding the use of the yolov8-rtdetr.yaml configuration file for training, it seems that there is a difference between YOLOv8 and RTDETR in terms of parsing the cfg file. The YOLOv8 code snippet you provided demonstrates how the cfg file is parsed using the overrides dictionary. However, in the RTDETR code snippet, the cfg file is not directly passed as a parameter for parsing.
> 
> It appears that the RTDETR module may not currently support parsing the ultra_tics/yolo/cfg/default.yaml file that is used for YOLOv8. It is likely that RTDETR has its own specific set of parameters that can be parsed.
> 
> To obtain more specific information on the parameters that can be parsed with RTDETR, I would recommend referring to the documentation or source code of RTDETR. These resources will provide you with detailed insights into the available configuration options and parameters for training your model.
> 
> I hope this explanation clarifies the situation for you. If you have any further questions or need additional assistance, please feel free to ask.
> 
> Best regards, Glenn Jocher

hi thank you for you reply！！
When I was using yaml for training, I encountered a problem where the model would initialize again.
it is my train in kaggle
!yolo train model=yolov8x-rtdetr.yaml data=coco8.yaml pretrained = yolov8x.pt
Subsequently, the model began to initialize

                   from  n    params  module                                       arguments                     
  0                  -1  1      2320  ultralytics.nn.modules.conv.Conv             [3, 80, 3, 2]                 
  1                  -1  1    115520  ultralytics.nn.modules.conv.Conv             [80, 160, 3, 2]               
  2                  -1  3    436800  ultralytics.nn.modules.block.C2f             [160, 160, 3, True]           
  3                  -1  1    461440  ultralytics.nn.modules.conv.Conv             [160, 320, 3, 2]              
  4                  -1  6   3281920  ultralytics.nn.modules.block.C2f             [320, 320, 6, True]           
  5                  -1  1   1844480  ultralytics.nn.modules.conv.Conv             [320, 640, 3, 2]              
  6                  -1  6  13117440  ultralytics.nn.modules.block.C2f             [640, 640, 6, True]           
  7                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
  8                  -1  3   6969600  ultralytics.nn.modules.block.C2f             [640, 640, 3, True]           
  9                  -1  1   1025920  ultralytics.nn.modules.block.SPPF            [640, 640, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  3   7379200  ultralytics.nn.modules.block.C2f             [1280, 640, 3]                
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  3   1948800  ultralytics.nn.modules.block.C2f             [960, 320, 3]                 
 16                  -1  1    922240  ultralytics.nn.modules.conv.Conv             [320, 320, 3, 2]              
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  3   7174400  ultralytics.nn.modules.block.C2f             [960, 640, 3]                 
 19                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  3   7379200  ultralytics.nn.modules.block.C2f             [1280, 640, 3]                
 22        [15, 18, 21]  1   7679244  ultralytics.nn.modules.head.RTDETRDecoder    [80, [320, 640, 640]]         
YOLOv8x-rtdetr summary: 485 layers, 67113884 parameters, 67113884 gradients

Downloading https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8x.pt to yolov8x.pt...
100%|█████████████████████████████████████████| 131M/131M [00:00<00:00, 191MB/s]
Transferred 510/725 items from pretrained weights

but，Later on, I discovered that the model was initialized again
![image](https://github.com/ultralytics/ultralytics/assets/115713675/5453dc82-f320-4309-b45f-dd6dfdd6e452)
And it seems that weight import has no effect
I checked the source code and it seems that the code that was continuously trained caused this problem
![image](https://github.com/ultralytics/ultralytics/assets/115713675/085a9c53-cfc2-4c88-bc86-0192cfc42a90)
if not overrides.get('resume'):  # manually set model only if not resuming
            self.trainer.model = self.trainer.get_model(weights=self.model if self.ckpt else None, cfg=self.model.yaml)It will reload and initialize the model again

Thank you very constructive.
your admirer:Bopeng

**Comment by glenn-jocher** [comment]
@liudongcheng hi Bopeng,

Thank you for providing more details about the issue you encountered while using the yolov8-rtdetr.yaml configuration file for training.

Based on the information you shared, it appears that the model is being initialized again during training and that the weight import does not seem to have any effect. This issue seems to be caused by the continuous training code in the source code.

The line of code you mentioned: 
`if not overrides.get('resume'):  # manually set model only if not resuming 
self.trainer.model = self.trainer.get_model(weights=self.model if self.ckpt else None, cfg=self.model.yaml)` 
seems to be responsible for reloading and initializing the model again.

To address this issue, you may need to modify this part of the code to prevent the model from being re-initialized during training. You can try commenting out or removing this code section or adjusting the conditions to ensure that the model is only loaded once.

I hope this explanation helps you understand the cause of the issue you are facing. If you have any further questions or need further assistance, please feel free to ask.

Best regards,
Glenn Jocher

**Comment by liudongcheng** [comment]
收到~

**Comment by glenn-jocher** [comment]
@liudongcheng thank you for your message. Please note that in the training process, model initialization might occur again due to certain conditions in the code. From your description, it seems this happening in your case. The loaded weights might not take effect depending on how the 'resume' parameter is set in the 'overrides' dictionary.

Essentially, the code checks if training is being resumed. If not, it resets the model and weights. Thus, to prevent the model from being re-initialized, you would need to ensure that the 'resume' parameter is set in your 'overrides' dictionary. This should signal the code to skip the initialization and utilization of pre-loaded weights.

I hope this clarifies your question. If you have further queries or need additional guidance, please feel free to ask.

**Comment by liudongcheng** [comment]
收到~

**Comment by liudongcheng** [comment]
收到~

**Comment by glenn-jocher** [comment]
@liudongcheng Bopeng,

很好，如有进一步问题或需求，请随时联系。

**Comment by liudongcheng** [comment]
收到~

**Comment by glenn-jocher** [comment]
@liudongcheng understood, thank you! If there's anything else you need assistance with, just let me know. 😊

**Comment by liudongcheng** [comment]
收到~

**Comment by glenn-jocher** [comment]
@liudongcheng 👍

**Comment by glenn-jocher** [comment]
@liudongcheng 收到您的消息！如果您有任何问题或需要帮助，请随时告诉我们。😊👍

**Comment by liudongcheng** [comment]
收到~

**Comment by glenn-jocher** [comment]
@liudongcheng 太好了，如果您有更多问题或需要进一步的帮助，请随时联系我们！😊

**Comment by liudongcheng** [comment]
收到~
